src/shared/libs/iota/quorum.js

```diff
@@ -0,0 +1,415 @@
+        }
+
+        return fallbackToSafeResult(method);
+    };
+};
+
+/**
+ *   Gets a safe result if quorum result falls below defined threshold
+ *
+ *   @method fallbackToSafeResult
+ *   @param {string} method
+ *
+ *   @returns {boolean}
+ **/
+const fallbackToSafeResult = (method) => {
+    const allowedMethodsMap = {
+        wereAddressesSpentFrom: true,
+        getInclusionStates: false,
+        getBalances: '0',
+        latestSolidSubtangleMilestone: EMPTY_HASH_TRYTES,
+    };
+
+    return allowedMethodsMap[method];
+};
+
+/**
+ *   For a list of nodes, find healthy nodes with size equals to minimum quorum nodes size
+ *
+ *   @method findSyncedNodes
+ *   @param {array} nodes
+ *   @param {array} [syncedNodes = []]
+ *   @param {array} [blacklistedNodes = []]
+ *
+ *   @returns {Promise}
+ **/
+const findSyncedNodes = (nodes, quorumSize, syncedNodes = [], blacklistedNodes = []) => {
+    const sizeOfSyncedNodes = size(syncedNodes);
+    const whitelistedNodes = filter(nodes, (node) => !includes(blacklistedNodes, node) && !includes(syncedNodes, node));
+
+    if (isEmpty(whitelistedNodes)) {
+        return Promise.reject(new Error(Errors.NOT_ENOUGH_HEALTHY_QUORUM_NODES));
+    }
+
+    const selectedNodes =
+        sizeOfSyncedNodes === quorumSize ? syncedNodes : sampleSize(whitelistedNodes, quorumSize - sizeOfSyncedNodes);
+
+    return Promise.all(map(selectedNodes, (provider) => isNodeSynced(provider).catch(() => undefined))).then(
+        (results) => {
+            const { activeNodes, inactiveNodes } = transform(
+                selectedNodes,
+                (acc, node, idx) => (results[idx] ? acc.activeNodes.push(node) : acc.inactiveNodes.push(node)),
+                { activeNodes: [], inactiveNodes: [] },
+            );
+
+            if (size(activeNodes) === size(selectedNodes)) {
+                return [...syncedNodes, ...activeNodes];
+            }
+
+            return findSyncedNodes(
+                nodes,
+                quorumSize,
+                // Add active nodes to synced nodes
+                [...syncedNodes, ...activeNodes],
+                // Add inactive nodes to blacklisted nodes
+                [...blacklistedNodes, ...inactiveNodes],
+            );
+        },
+    );
+};
+
+/**
+ *   From a list of synced nodes, compute a quorum result for wereAddressesSpentFrom iota api
+ *
+ *   @method getQuorumForWereAddressesSpentFrom
+ *   @param {array} payload - Addresses
+ *   @param {array} syncedNodes
+ *
+ *   @returns {Promise}
+ **/
+const getQuorumForWereAddressesSpentFrom = (payload, syncedNodes) => {
+    const requestPayloadSize = size(payload);
+
+    return rejectIfNotEnoughSyncedNodes(syncedNodes)
+        .then(() =>
+            Promise.all(
+                map(
+                    syncedNodes,
+                    (provider) =>
+                        new Promise((resolve) => {
+                            new IOTA({ provider }).api.wereAddressesSpentFrom(
```

Comment:
does this have some type of timeout? so err (or valid response) is promised to be returned eventually?

Reply:
It will eventually timeout (because of the addition [here)](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L272) and an error or a valid response is guaranteed.